File: examples/python/receive_commands_noserial.py
import cobs
import cobs.cobs
import rscp_protobuf
import io
import time

# This simulated buffer contains 
# raw bytes that represent a message 
# that was receiver over serial module
# to simulate RSCP Host -> Rover communication
simulated_buffer = io.BytesIO(b'\x05\n\x02\x08\x01\x00\x05\x12\x02\x08\x03\x00\x1c\x1a\x19\n\x17\tX9\xb4\xc8v\xbe\xf3?\x11\x83\xc0\xca\xa1E\xb6\x16@\x1d\xa1\xd6|?\x00\x18"\x19\n\x12\tX9\xb4\xc8v\xbe\xf3?\x11\x83\xc0\xca\xa1E\xb6\x16@\x15\x01\x03\xc8B\x00')

def on_receive(data: bytes):
    # first step is to COBS decode the data
    cobs_decoded = cobs.cobs.decode(data)
    
    request = rscp_protobuf.RequestEnvelope()
    # next step is to decode(parse) the protobuf message
    request.ParseFromString(cobs_decoded)
    
    print(f"Received new message at: {time.time()}")
    print(f"Received Request type: {request.WhichOneof('request')}")
    print(request)

if __name__ == "__main__":
    buffer = b""
    while True:
        # instead of this, you may use serial.read() to read from serial port
        # see receive_commands.py for serial example
        data = simulated_buffer.read(1)
        if data == b"\x00":
            on_receive(buffer)
            buffer = b""
        else:
            buffer += data
        

File: examples/Cpp/receive_commands_noserial.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include "rscp.pb.h"      // Generated protobuf header
#include "Include\cobs.h" // COBS (Consistent Overhead Byte Stuffing) encoding/decoding

/**
 * @brief Callback function to handle and decode incoming COBS-encoded protobuf messages.
 *
 * This function takes a byte buffer representing a COBS-encoded Protocol Buffers message,
 * decodes it using the COBS algorithm, and attempts to deserialize it into a
 * `rscp::RequestEnvelope` object. If successful, it prints out the request type and
 * a full debug string of the message contents.
 *
 * @param buffer A reference to a vector of bytes representing the received data frame.
 *
 * @note If decoding or parsing fails, an error message is printed to stderr.
 */
void on_receive(const std::vector<uint8_t> &buffer)
{
    // Decode the received buffer using COBS
    auto decoded = cobs_decode(buffer.data(), buffer.size());

    // Deserialize the protobuf message into a RequestEnvelope object
    rscp::RequestEnvelope request;
    if (!request.ParseFromArray(decoded.data(), decoded.size()))
    {
        std::cerr << "Failed to parse cmakprotobuf message.\n";
        return;
    }

    // Log the received request type and full debug output
    std::cout << "Received Request type: " << request.request_case() << std::endl;
    std::cout << request.DebugString() << std::endl;
}

/**
 * main function
 */
int main()
{
    rscp::RequestEnvelope request;

    // Simulated raw COBS-encoded data stream (might represent multiple frames)
    std::vector<uint8_t> simulated_data = {
        0x05, 0x0A, 0x02, 0x08, 0x01, 0x00, 0x05, 0x12, 0x02, 0x08, 0x03, 0x00, 0x1C, 0x1A, 0x19, 0x0A, 0x17,
        0x09, 0x58, 0x39, 0xB4, 0xC8, 0x76, 0xBE, 0xF3, 0x3F, 0x11, 0x83, 0xC0, 0xCA, 0xA1, 0x45, 0xB6, 0x16,
        0x40, 0x1D, 0xA1, 0xD6, 0x7C, 0x3F, 0x00, 0x18, 0x22, 0x19, 0x0A, 0x12, 0x09, 0x58, 0x39, 0xB4, 0xC8,
        0x76, 0xBE, 0xF3, 0x3F, 0x11, 0x83, 0xC0, 0xCA, 0xA1, 0x45, 0xB6, 0x16, 0x40, 0x15, 0x01, 0x03, 0xC8,
        0x42, 0x00};

    std::vector<uint8_t> buffer;

    // Split the stream into frames using 0x00 as a frame delimiter
    for (auto byte : simulated_data)
    {
        if (byte == 0x00)
        {
            // Process the complete frame
            on_receive(buffer);
            buffer.clear();
        }
        else
        {
            // Accumulate bytes into the current frame
            buffer.push_back(byte);
        }
    }

    return 0;
}
